packages/dds/shared-object-base/src/sharedObject.ts

```diff
@@ -450,3 +364,89 @@ export abstract class SharedObject<TEvent extends ISharedObjectEvents = ISharedO
 
     protected abstract applyStashedOp(content: any): unknown;
 }
+
+/**
+ *  Base class for shared objects with additional functionality for synchronous summarization, serializer, and GC
+ */
+export abstract class SharedObject<TEvent extends ISharedObjectEvents = ISharedObjectEvents>
+    extends SharedObjectCore<TEvent> {
+    /**
+     * True while we are capturing state to prevent use of regular serializer.
+     */
+    private _isCapturing: boolean = false;
+
+    /**
+     * The serializer to use to serialize / parse handles, if any.
+     */
+    private readonly _serializer: IFluidSerializer;
+
+    protected get serializer(): IFluidSerializer {
+        /**
+         * During revision capture, only SummarySerializer should be used (instead of this.serializer) to keep track of
+         * IFluidHandles that are serialized for GC. These handles represent references to other Fluid objects.
+         */
+        assert(!this._isCapturing,
+            0x075 /* "SummarySerializer should be used for serializing data during summary." */);
+        return this._serializer;
+    }
+
+    /**
+     * @param id - The id of the shared object
+     * @param runtime - The IFluidDataStoreRuntime which contains the shared object
+     * @param attributes - Attributes of the shared object
+     */
+    constructor(
+        id: string,
+        runtime: IFluidDataStoreRuntime,
+        attributes: IChannelAttributes)
+    {
+        super(id, runtime, attributes);
+
+        this._serializer = new FluidSerializer(this.runtime.channelsRoutingContext);
+    }
+
+    /**
+     * {@inheritDoc (ISharedObject:interface).captureRevision}
+     */
+    public captureRevision(): IChannelRevision {
```

Comment:
I see a couple of problems with having this API instead of separate APIs for summarize and getGCData:
1. How will a DDS have custom implementation for one of these APIs? For example, SharedMatrix and SharedString work on a different set of data for generating GC state vs summary state. (Btw, this change will break them).
2. The state returned by summarize is affected by GC. When GC runs on the data returned by getGCData, it may update the "reference" state of certain nodes in the system. This information needs to be encoded in the summary. Additionally, some nodes which did not need summarization before (because its data did not change) may need to be summarized because its "reference" state changed.
This is not really a problem at this layer yet but as soon as we have sub-DDS GC, this will break. Also, IFluidDataStoreChannel's shape will also become similar to this and it will run into this problem now.

Reply:
The main challenge with point 2 above is that we don't fully know whether we need to re-summarize a node or not until GC has run.

Reply:
1. I'm not sure if my changes for sequence and matrix are the right thing. For DDSs that want to do something custom, they should derive from SharedObjectCore and fully implement captureRevision themselves. They'd still need to capture all the state necessary for both summaries and GC.
2. Is it possible to update the captured state and generate another summary off that? or does it need to happen on the sharedObject?

Reply:
I was asking myself that question (that Navin raised) - can summary and gc data come from same state or not, sounds not (sorry I forgot it). I think there are two ways to address it:
- Leave API as is but capture revision twice (at different times).
- Leave GC API the same, while revision to support only summary API.

It's possible that the answer will be different on the layer. I.e. the general structure should likely allow flexibility, and this we might want to leave GC API as is (including on SharedObjectCore layer). But if all but Whiteboard DDSs can always generage GC from summary, then I'd argue we want to build code in a way where this duplication of logic happens in single place (SharedObject) by having implementation of IChannelRevision here that can extract GC data out of it (i.e. have concrete implementation of IChannelRevision that captures and exposes GC data). That way, SharedObject.getGCData() can just do this.captureRevision().getGcData().

Reply:
I don't think we can implement it via the first way you mentioned. The reason is that the summary state of a node may have the reference state (whether its unreferenced or not) based on an older summary state. Basically, GC state can be from seq#100 whereas summary state can be from seq#200. We cannot mark a node at seq#200 referenced / unreferenced based on its data from seq#100 - that would be inconsistent.

Both the summary and GC state in a summary should be off the same point in time (seq#).

Reply:
I'm not sure how to interpret your comment :) DDS just exposes ability to capture these states, at any moment in time. External process governs when these methods are called - for same state or not. DDS itself does not capture any additional state, so there is no differences here. The only thing that differs is how internally it captures GC - if all SharedObject derived classes extract GC data in exactly same way (by essentially summarizing themselves), then there is no need to duplicate this code - SharedObject base class can help here.

Reply:
I misunderstood this change. I was under the impression that we are also changing the way summarization happens in the regular case in summarizer client.